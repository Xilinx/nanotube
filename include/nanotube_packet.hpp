/**************************************************************************\
*//*! \file nanotube_packet.hpp
** \author  Stephan Diestelhorst <stephand@amd.com>
**          Neil Turton <neilt@amd.com>
**  \brief  Nanotube packet internal definitions.
**   \date  2020-08-26
*//*
\**************************************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/

#ifndef NANOTUBE_PACKET_HPP
#define NANOTUBE_PACKET_HPP

#include <cstddef>
#include <cstdint>
#include <memory>
#include <vector>

#include "nanotube_api.h"
#include "bus_id.h"

// Defines sections of a packet.  The interpretation of the sections
// may depend on the bus type.
// For the time being we only support WHOLE, METADATA, and PAYLOAD.
typedef enum {
  // The whole packet, including all metadata.
  NANOTUBE_SECTION_WHOLE,

  // The packet payload.  Typically Ethernet.
  NANOTUBE_SECTION_PAYLOAD,

  // All the metadata.  This may be the same as
  // NANOTUBE_SECTION_SYS_META.
  NANOTUBE_SECTION_METADATA,

  // The system metadata.  This may or may not include
  // NANOTUBE_SECTION_USER_META.
  NANOTUBE_SECTION_SYS_META,

  // The user metadata, available for application use.
  NANOTUBE_SECTION_USER_META,
} nanotube_packet_section_t;

/*!
** Implementation of a Nanotube packet.
**
** Most users should not have to care about the specific contents / layout,
** here!
*/
class nanotube_packet {
public:
  /*! Construct a packet. */
  nanotube_packet(enum nanotube_bus_id_t bus_type = NANOTUBE_BUS_ID_ETH) {
    reset(bus_type);
  }

  /*! Reset the packet contents with the specified bus type. */
  void reset(enum nanotube_bus_id_t bus_type);

  /*! Reset the packet contents without changing the bus type. */
  void reset() { reset(m_bus_type); }

  enum nanotube_bus_id_t get_bus_type() const {
    return m_bus_type;
  }

  /*! Get the flag indicating whether the packet should be treated as
   *  a capsule. */
  bool get_is_capsule() const { return m_is_capsule; }

  /*! Set the flag indicating whether the packet should be treated as
   *  a capsule. */
  void set_is_capsule(bool val) { m_is_capsule = val; }

  /*! Convert the packet to a different bus type.
   *
   * \param bus_type The target bus type.
   * \returns a Linux error code.
   */
  int convert_bus_type(enum nanotube_bus_id_t bus_type);

  /*! Get a pointer to the beginning of a section. */
  const uint8_t *begin(nanotube_packet_section_t sec) const {
    /* Call the non-const method and cast the result.  Use of a
     * non-const temporary makes it clear that this method isn't
     * calling itself recursively. */
    uint8_t *res = const_cast<nanotube_packet *>(this)->begin(sec);
    return res;
  }

  /*! Get a pointer to the beginning of a section. */
  uint8_t *begin(nanotube_packet_section_t sec);

  /*! Get a pointer to the end of a section. */
  const uint8_t *end(nanotube_packet_section_t sec) const {
    /* Call the non-const method and cast the result.  Use of a
     * non-const temporary makes it clear that this method isn't
     * calling itself recursively. */
    uint8_t *res = const_cast<nanotube_packet *>(this)->end(sec);
    return res;
  }

  /*! Get a pointer to the end of a section. */
  uint8_t *end(nanotube_packet_section_t sec);

  /*! Get the size of a section. */
  std::size_t size(nanotube_packet_section_t sec) const;

  /*! Read bytes from a section. */
  void read(nanotube_packet_section_t sec, uint8_t *buffer,
            std::size_t offset, std::size_t length);

  /*! Write bytes into a section. */
  void write(nanotube_packet_section_t sec, const uint8_t *buffer,
             std::size_t offset, std::size_t length);

  /*! Set the size of a section.  The section is truncated or filled
   *  with zeros. Only support SECTION_WHOLE for now. */
  void resize(nanotube_packet_section_t sec, std::size_t size);

  /*! Adjust the size of the section.  Invalidates all pointers to
   *  data in this packet.
   *
   * \param sec     The section to adjust.
   * \param offset  The offset at which to insert/remove bytes.
   * \param adjust  The number of bytes inserted (positive) into or
   *                removed (negative) from the packet.
   */
  void resize(nanotube_packet_section_t sec, std::size_t offset,
              int32_t adjustment);

  /*! Insert bytes into a section. */
  void insert(nanotube_packet_section_t sec, const uint8_t *buffer,
              std::size_t offset, std::size_t length);

  /*! Get the current port number. */
  nanotube_packet_port_t get_port() const;

  /*! Set the current port number. */
  void set_port(nanotube_packet_port_t port);

  /*! Get a bus word from the packet.
   *
   * \param buffer   The buffer to receive the word.
   * \param buf_size The size of the buffer in bytes.
   * \param iter     A pointer to an iterator.
   *
   * \returns true if there are more words.
   *
   * The buffer size must be correct for the bus type.  The value of
   * iter on the first call must be zero and it will be updated for
   * the next call.
   */
  bool get_bus_word(uint8_t *buffer, std::size_t buf_size,
                    std::size_t *iter);

  /*! Add a bus word to the packet.
   *
   * \param buffer   The buffer to receive the word.
   * \param buf_size The size of the buffer in bytes.
   *
   * \returns true if there are more words.
   *
   * The buffer size must be correct for the bus type.  The packet
   * should be reset before calling this method for the first word.
   * The packet will be valid after this method returns true.
   */
  bool add_bus_word(uint8_t *buffer, std::size_t buf_size);

private:
  /*! Get the size of the metadata header. */
  std::size_t get_meta_size() const;

  /*! The bus type used to represent the packet. */
  enum nanotube_bus_id_t m_bus_type;

  /*! A flag indicating whether the packet should be treated as a
   *  capsule.  This is only used for the C API. */
  bool m_is_capsule;

  /*! The contents of the packet. Contains metadata prefixed to packet data. */
  std::vector<uint8_t> m_contents;

  /*! The current destination port of the packet. */
  nanotube_packet_port_t m_port;

  /*! Byte offset to start of packet data in contents. */
  std::size_t m_meta_size;

  bool m_data_eop_seen; // Only used for x3rx bus EOP tracking
};

typedef std::unique_ptr<struct nanotube_packet> nanotube_packet_ptr_t;

#endif // NANOTUBE_PACKET_HPP
