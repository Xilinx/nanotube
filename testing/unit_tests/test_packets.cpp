/*******************************************************/
/*! \file test_packets.cpp
** \author Stephan Diestelhorst <stephand@amd.com>
**  \brief Unit tests for Nanotube packet implementations.
**   \date 2020-03-04
*//******************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "nanotube_packet.hpp"
#include "nanotube_api.h"

static void print_bytes(uint8_t* data, size_t len) {
  for( size_t i = 0; i < len; ++i ) {
    printf("%02x ", data[i]);
    if( (i % 8) == 7 )
      printf("\n");
  }
}


/*!
** Allocate a new packet of specified size.
** \param len Size of the packet in bytes.
** \return Pointer to a freshly allocated packet.
**/
nanotube_packet_t *malloc_pkt(size_t len) {
  nanotube_packet_t *pkt = new nanotube_packet_t();
  auto sec = NANOTUBE_SECTION_WHOLE;
  pkt->resize(sec, len);
  return pkt;
}

/*!
** Free a Nanotube packet.
** \param pkt Pointer to the packet that will be freed.
**/
void free_pkt(nanotube_packet_t *pkt) {
  delete pkt;
}

/*!
** Test Nanotube packet access functions.
** \param len Packet size to perform the tests with
** \return Number of errors.
**/
int test_packets(size_t pkt_len) {
  int err = 0;
  const size_t buffer_size = 16384;
  printf("\n\n_Testing_Packets_\n");
  nanotube_packet_t *pkt = malloc_pkt(pkt_len);

  /* Basics, first :) */
  size_t len;
  len = nanotube_packet_bounded_length(pkt, 16);
  printf("Packet restricted length %zu\n", len);
  if( len != 16 )
    ++err;

  len = nanotube_packet_bounded_length(pkt, buffer_size);
  printf("Packet length %zu\n", len);
  if( len != pkt_len && pkt_len < buffer_size )
    ++err;

  auto sec = NANOTUBE_SECTION_WHOLE;
  assert(nanotube_packet_data(pkt) == pkt->begin(sec));

  uint8_t *data = (uint8_t*)malloc(buffer_size);

  /* Fill packet with test data */
  auto pkt_data = pkt->begin(sec);
  for( size_t i = 0; i < pkt_len; ++i )
    pkt_data[i] = i & 0xff;

  /* Read / write x offsets & lengths */
  size_t n;
  n = nanotube_packet_read(pkt, data, 0, len);
  printf("Read %zu bytes, asking for %zu\n", n, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;

  size_t offs = 16;
  n = nanotube_packet_read(pkt, data, offs, 8);
  printf("Read %zu bytes from offset %zu, asking for %i\n", n, offs, 8);
  print_bytes(data, len); printf("\n");
  if( n != 8 )
    ++err;
  if( data[0] != data[offs] )
    ++err;

  /* Small write + offset */
  uint8_t *all_enabled = (uint8_t *)malloc(buffer_size / 8);
  memset(all_enabled, 0xff, buffer_size / 8);
  uint8_t write_data[] = {0xde, 0xad, 0xbe, 0xef};

  offs = 40;
  n = nanotube_packet_write_masked(pkt, write_data, all_enabled, offs,
                                   sizeof(write_data));
  printf("Wrote %zu bytes to offset %zu, with req size %zu\n", n, offs,
         sizeof(write_data));
  if( n != sizeof(write_data) )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if ( memcmp(data + offs, write_data, 4) != 0 )
    ++err;

  /* Small unmasked write */
  offs = 44;
  n = nanotube_packet_write(pkt, write_data, offs, sizeof(write_data));
  printf("Wrote %zu bytes to offset %zu, with req size %zu\n", n, offs,
         sizeof(write_data));
  if( n != sizeof(write_data) )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if ( memcmp(data + offs, write_data, 4) != 0 )
    ++err;

  /* Full write */
  for( size_t i = 0; i < len; ++i )
    data[i] ^= 0x0f;
  n = nanotube_packet_write_masked(pkt, data, all_enabled, 0, len);
  printf("Wrote %zu bytes, with req size %zu\n", n, len);
  if( n != len )
    ++err;
  memset(data, 0, len);
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;

  /* Write w/ enables */
  uint8_t *fancy_enable = (uint8_t*)malloc(len / 8);
  memset(fancy_enable, 0, len / 8);
  {
    /* Craft a pattern of alternating on/off that are getting larger */
    size_t i = 0;
    size_t s = 1;
    while ( i < len ) {
      /* Skip S bits */
      i = i + s;
      /* Fill S bits */
      for( size_t j = 0; j < s && i < len; ++i, ++j )
        fancy_enable[i / 8] |= (1 << (i % 8));
      s = s * 2;
    }
  }
  memset(data, 0xff, len);
  offs = 16;
  n = nanotube_packet_write_masked(pkt, data, fancy_enable, offs,
                   len - offs);
  printf("Wrote %zu bytes, with fancy mask, offset %zu, req size %zu\n", n,
         offs, len - offs);
  if( n != len - offs )
    ++err;
  memset(data, 0, len);
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( data[offs + 1] != 0xff || data[offs + 4] != 0xff ||
      data[offs + 5] != 0xff )
    ++err;

  /* Corner cases: Reads */
  /* Outside packet */
  offs = len + 8;
  n = nanotube_packet_read(pkt, data, offs, len);
  printf("Read %zu bytes, asking for %zu from offset %zu\n", n, len, offs);
  print_bytes(data, len); printf("\n");
  if( n != 0 )
    ++err;

  /* Partial & across end of packet */
  offs = len - 8;
  n = nanotube_packet_read(pkt, data, offs, len);
  printf("Read %zu bytes, asking for %zu from offset %zu\n", n, len, offs);
  print_bytes(data, len); printf("\n");
  if( n != len - offs )
    ++err;
  if( data[0] != 0x37 || data[n] != 0 )
    ++err;

  /* Empty read */
  offs = 0;
  n = nanotube_packet_read(pkt, data, offs, 0);
  printf("Read %zu bytes, asking for %i\n", n, 0);
  if( n != 0 )
    ++err;

  /*print_bytes(data, len); printf("\n"); */
  /* Read from NULL packet */
  n = nanotube_packet_read(NULL, data, offs, len);
  printf("Read %zu bytes, asking for %zu from NULL packet\n", n, len);
  print_bytes(data, len); printf("\n");
  if( n != 0 )
    ++err;

  /* Read to NULL data */
  n = nanotube_packet_read(pkt, NULL, offs, len);
  printf("Read %zu bytes, asking for %zu to NULL data array\n", n, len);
  if( n != 0 )
    ++err;

  /* Corner cases: Writes */
  /* Outside packet */
  offs = len + 8;
  n = nanotube_packet_write_masked(pkt, write_data, all_enabled, offs,
                   sizeof(write_data));
  printf("Wrote %zu bytes to offset %zu, with req size %zu\n", n, offs,
      sizeof(write_data));
  if( n != 0 )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;

  n = nanotube_packet_write(pkt, write_data, offs, sizeof(write_data));
  printf("Wrote %zu bytes to offset %zu, with req size %zu\n", n, offs,
      sizeof(write_data));
  if( n != 0 )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;

  /* Partial & across end of packet */
  offs = len - sizeof(write_data) / 2;
  n = nanotube_packet_write_masked(pkt, write_data, all_enabled, offs,
                   sizeof(write_data));
  printf("Wrote %zu bytes to offset %zu, with req size %zu\n", n, offs,
         sizeof(write_data));
  if( n != len - offs )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;
  if( memcmp(data + offs, write_data, len - offs) != 0 )
    ++err;

  uint8_t write_data2[] = {0xbe, 0xda, 0x22, 0x1e};
  n = nanotube_packet_write(pkt, write_data2, offs, sizeof(write_data));
  printf("Wrote %zu bytes to offset %zu, with req size %zu\n", n, offs,
         sizeof(write_data2));
  if( n != len - offs )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;
  if( memcmp(data + offs, write_data2, len - offs) != 0 )
    ++err;

  /* Empty write */
  offs = 0;
  n = nanotube_packet_write_masked(pkt, write_data, all_enabled, offs, 0);
  printf("Wrote %zu bytes to offset %zu, with req size %i\n", n, offs, 0);
  if( n != 0 )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;

  n = nanotube_packet_write(pkt, write_data, offs, 0);
  printf("Wrote %zu bytes to offset %zu, with req size %i\n", n, offs, 0);
  if( n != 0 )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;

  /* Write to NULL packet */
  n = nanotube_packet_write_masked(NULL, write_data, all_enabled, offs,
                   sizeof(write_data));
  printf("Wrote %zu bytes to NULL packet\n", n);
  if( n != 0 )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;

  n = nanotube_packet_write(NULL, write_data, offs, sizeof(write_data));
  printf("Wrote %zu bytes to NULL packet\n", n);
  if( n != 0 )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;

  /* Write with NULL data */
  n = nanotube_packet_write_masked(pkt, NULL, all_enabled, offs,
                   sizeof(write_data));
  printf("Wrote %zu bytes with NULL data\n", n);
  if( n != 0 )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;

  n = nanotube_packet_write(pkt, NULL, offs, sizeof(write_data));
  printf("Wrote %zu bytes with NULL data\n", n);
  if( n != 0 )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;

  /* Write with NULL enables */
  n = nanotube_packet_write_masked(pkt, write_data, NULL, offs,
                   sizeof(write_data));
  printf("Wrote %zu bytes with NULL enables\n", n);
  if( n != 0 )
    ++err;
  n = nanotube_packet_read(pkt, data, 0, len);
  print_bytes(data, len); printf("\n");
  if( n != len )
    ++err;


  free(all_enabled);
  free(data);
  free(fancy_enable);
  free_pkt(pkt);

  return err;
}


int main(int argc, char *argv[]) {
  int num_errs = 0;

  num_errs += test_packets(64);
  /*test_packets(1500); */
  /*test_packets(9000); */

  printf("Found %i errors\n", num_errs);
  return num_errs;
}
/* vim: set ts=8 et sw=2 sts=2 tw=75: */
