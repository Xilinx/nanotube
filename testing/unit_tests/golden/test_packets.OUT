

_Testing_Packets_
Packet restricted length 16
Packet length 64
Read 64 bytes, asking for 64
00 01 02 03 04 05 06 07 
08 09 0a 0b 0c 0d 0e 0f 
10 11 12 13 14 15 16 17 
18 19 1a 1b 1c 1d 1e 1f 
20 21 22 23 24 25 26 27 
28 29 2a 2b 2c 2d 2e 2f 
30 31 32 33 34 35 36 37 
38 39 3a 3b 3c 3d 3e 3f 

Read 8 bytes from offset 16, asking for 8
10 11 12 13 14 15 16 17 
08 09 0a 0b 0c 0d 0e 0f 
10 11 12 13 14 15 16 17 
18 19 1a 1b 1c 1d 1e 1f 
20 21 22 23 24 25 26 27 
28 29 2a 2b 2c 2d 2e 2f 
30 31 32 33 34 35 36 37 
38 39 3a 3b 3c 3d 3e 3f 

Wrote 4 bytes to offset 40, with req size 4
00 01 02 03 04 05 06 07 
08 09 0a 0b 0c 0d 0e 0f 
10 11 12 13 14 15 16 17 
18 19 1a 1b 1c 1d 1e 1f 
20 21 22 23 24 25 26 27 
de ad be ef 2c 2d 2e 2f 
30 31 32 33 34 35 36 37 
38 39 3a 3b 3c 3d 3e 3f 

Wrote 4 bytes to offset 44, with req size 4
00 01 02 03 04 05 06 07 
08 09 0a 0b 0c 0d 0e 0f 
10 11 12 13 14 15 16 17 
18 19 1a 1b 1c 1d 1e 1f 
20 21 22 23 24 25 26 27 
de ad be ef de ad be ef 
30 31 32 33 34 35 36 37 
38 39 3a 3b 3c 3d 3e 3f 

Wrote 64 bytes, with req size 64
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f 1e 1d 1c 1b 1a 19 18 
17 16 15 14 13 12 11 10 
2f 2e 2d 2c 2b 2a 29 28 
d1 a2 b1 e0 d1 a2 b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 31 30 

Wrote 48 bytes, with fancy mask, offset 16, req size 48
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 ff ff 

Read 0 bytes, asking for 64 from offset 72
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 

Read 8 bytes, asking for 64 from offset 56
37 36 35 34 33 32 ff ff 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 

Read 0 bytes, asking for 0
Read 0 bytes, asking for 64 from NULL packet
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 
00 00 00 00 00 00 00 00 

Read 0 bytes, asking for 64 to NULL data array
Wrote 0 bytes to offset 72, with req size 4
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 ff ff 

Wrote 0 bytes to offset 72, with req size 4
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 ff ff 

Wrote 2 bytes to offset 62, with req size 4
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 de ad 

Wrote 2 bytes to offset 62, with req size 4
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 be da 

Wrote 0 bytes to offset 0, with req size 0
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 be da 

Wrote 0 bytes to offset 0, with req size 0
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 be da 

Wrote 0 bytes to NULL packet
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 be da 

Wrote 0 bytes to NULL packet
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 be da 

Wrote 0 bytes with NULL data
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 be da 

Wrote 0 bytes with NULL data
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 be da 

Wrote 0 bytes with NULL enables
0f 0e 0d 0c 0b 0a 09 08 
07 06 05 04 03 02 01 00 
1f ff 1d 1c ff ff 19 18 
17 16 ff ff ff ff 11 10 
2f 2e 2d 2c 2b 2a ff ff 
ff ff ff ff ff ff b1 e0 
3f 3e 3d 3c 3b 3a 39 38 
37 36 35 34 33 32 be da 

Found 0 errors
